### PR TITLE
Twig Rendering Service + React Schema Form + Resizable Component Viewer 

### PR DESCRIPTION
**Pull request by sghoweri**
image

<hr>

TLDR: this is the first of many architecture-level updates that'll be integral to Bolt v2 and beyond.

I know this PR isn’t perfect (caveats below), but as an MVP this PR sets us up for some truly next gen stuff already in progress 😉.

# Summary

Besides providing component's with a dynamic, schema-powered Form that makes requests to the new Twig Rendering Service, including a resizable iframe UI that all renders to, this work will help unlock so many new exciting changes for Bolt in the near future including:
- Component-level API and snapshot testing via Jest
- Unified Twig rendering of the docs + Pattern Lab sites
- Much much **much** faster Twig compiles
- Dynamically generated / customized / personalized Pattern Lab / docs
- PHP-based Webpack Twig Loader
- The eventual migration away from Pattern Lab PHP and onto Pattern Lab Node

# Work Done So Far

## Updated Preact config to support even more in the React ecosystem (including React DOM)

- Updated build tools to support using the majority of the off the shelf React components out there with our current set up (still using Preact under the hood) via `preact-compat`

## Eslint updates

- Updated Eslint config to manually use the same package alias data for Preact / React so linting doesn’t complain about React and React DOM missing
- Addressed Eslint Issues with the schema form-related updates

### Webpack Serve
- `webpack-dev-server` replaced with `webpack-serve`
  - Matches CLI message improvements added in on the epic/v2 branch
  - Fixed HMR issues so BrowserSync + Webpack should auto reload more predictably now

### More Resilient Build Task Config
- Adjusted how individual build tasks grab config data — more tasks can now be run independently, even outside of the Bolt CLI interface
  - For example, Webpack’s CLI tool can now be run from the command line using our Webpack config (super helpful for testing the different pieces of the build process independently)

## Component Rendering Service 
- Updated build to support spinning up a PHP server + middleware for serving up our component rendering service
- Updated build config to assign a unique, open port to the new renderingService
- Updated PL to only render schema form when renderingService is enabled
- Reorganized + updated rendering service to support being used from PL + by Node

## Component Explorer + Schema Editor
- Added the new internal React JSON schema form component to PL
- Updated the (new internal-only) schema form component to render the HTML from the rendering service into a dynamically created iframe
  - **NOTE:** this technique being tested out here will also provide us with a way to render out component previews in Drupal that are fully encapsulated!
- Added custom UI + functionality to make the iframe container resizable (additional work needed on this but the basics work as is)
- Began cleaning up styles of the out of the box JSON schema form component
  - Significantly consolidate UI to take up less vertical space
  - Hide empty fields
  - Added missing checkbox styles
- Added localStorage caching to save layout resizing preferences
  - Added a "Reset Layout" button control to wipe cached data for resized layout and reset to default
- Added localStorage caching + reset to cache rendered HTML response + custom form data
  - Also added a "Reset Form" button control to wipe form data back to initial defaults

<hr>

Navbar docs w/ the PHP Rendering Service (left) vs the same Navbar docs page when compiled via Travis (right) which automatically disables the rendering service config option till we have a docker container up on Now.sh that can run the PHP scripts required here:

image

## Caveats
- the CSS and JS for the schema form are rough around the edges at best (seriously, they're in place to make this feature look OK and demoable but this schema form / viewer code isn't something we'd be shipping as-is)
- The new Twig Rendering Service here _only_ works locally for now. Getting this to work up on Now.sh support will require us to get a Docker container up and running which I’d prefer we tackle separately
- Only the Navbar docs are wired up to display the new schema form / component preview UI. The rest of the components docs in PL will need to be configured slightly to start using this new feature (again, separate update)
- Once we merge this in I’ll be swapping out this initial Twig Rendering Service php logic out with the vastly improved version Evan and I (but mostly Evan) has been working on: https://github.com/basaltinc/twig-renderer

### CODE REVIEW

File: packages/core-php/src/TwigExtensions/BoltExtras.php [hunk 1]
```diff
@@ -28,6 +28,8 @@ public function getFilters() {
       Bolt\TwigFilters::markdown(),
       Bolt\TwigFilters::rgb2hex(),
       Bolt\TwigFilters::text_contrast(),
+      Bolt\TwigFilters::without(),
+      Bolt\TwigFilters::t(),
```
Comment (remydenton) [inline packages/core-php/src/TwigExtensions/BoltExtras.php]: I can't for the life of me figure out where these were coming from before 🤔 .  Could you briefly explain (mostly for my benefit) where they got removed from (or will get removed from)?

File: packages/build-tools/utils/config.schema.yml [hunk 2]
```diff
@@ -45,6 +45,15 @@
         - pl
         - static
         - drupal
+        - hybrid
```
Comment (remydenton) [inline packages/build-tools/utils/config.schema.yml]: I think the meanings of `pl`, `static`, and `drupal` are all intuitive, whereas `hybrid` isn't.  I'd like to see a brief bit of documentation to explain it; I recommend the `description` field here in the schema as an appropriate place.

**Comment by sghoweri** [comment]
👍🏼 great questions!

File: .eslintignore [hunk 3]
```diff
@@ -4,6 +4,7 @@
 /www/**
 /apps/drupal-lab/**
 /apps/pw-site/www/**
+/apps/pattern-lab/src/components/schema-form/**
```
[the remark this answers is not in the document]
Comment (sghoweri) [reply .eslintignore]: I agree. This wasn’t intended to be a long term fix — let me take another crack at buttoning this up so this opt out won’t be needed.

[the remark this answers is not in the document]
Comment (sghoweri) [reply .eslintignore]: All set with this one!

**Comment by joekarasek** [comment]
React-json-schema-form WTF!!!
Looks great!

File: packages/core-php/src/TwigExtensions/BoltExtras.php (hunk 1 above)
Comment (sghoweri) [reply packages/core-php/src/TwigExtensions/BoltExtras.php]: @remydenton for now these aren't literally getting removed.

For reference however, the original versions of these came from here: https://github.com/pattern-lab/drupal-twig-extensions. Long story short, "pretending" to be the same Twig extension as what's in Drupal is sometimes very different than literally sharing the same logic, hence why we've added our own take on a lot of these (the without filter being an excellent example of this) since we can get a 1:1 match with how Drupal does a lot of these things under the hood.

Back to this change specifically... while this PR isn't removing that Pattern Lab Drupal Twig extensions repo, it'll likely be going away at some point in the near future. The updates I'm working on locally (combining Pattern Lab with the Docs site) ran into some snags when wiring up the unified Twig rendering engine to compile everything across the board. 

This update here, while harmless to the outside world -- `core-php` splits up the shared Twig extensions into `BoltExtras` which is specific to us vs `BoltCore` which is shared with Drupal -- .preps us for those larger updates to come 😉.

**Comment by sghoweri** [comment]
Thanks @joekarasek - miss ya buddy!

So much crazy cool stuff brewing!! ^_^

File: packages/build-tools/utils/config.schema.yml (hunk 2 above)
Comment (sghoweri) [reply packages/build-tools/utils/config.schema.yml]: @remydenton I've updated the description on this new environmental config option (again, prepping for in-flight work) and renamed this to instead be `pwa` standing for "Progressive Web App" since that's literally what this new unified PL + Docs hybrid is shaping up to be!

**Comment by sghoweri** [comment]
@remydenton I've added comments / pushed out updates based on all your original comments btw! 

I know some of the under the hood async build task updates have been added since then -- just wanted to do a gut check to see what else was needed to get this ready to merge down so we can keep iterating on this.

**Comment by sghoweri** [comment]
Excellent - thanks @remydenton!
